### DISCUSSION TITLE
Recommendation IDE LLDB debugger

### DISCUSSION THREAD

**Discussion by Blargian** [ClickHouse/ClickHouse#59673] [General]
I've been using Visual Studio based on recommendation of @kitaisreal in this video [Guide to your first contribution to ClickHouse](https://youtu.be/tNzdrw1L1Xk?si=udLDDgvvNEnMP7mJ) along with extension [Code LLDB](https://github.com/vadimcn/codelldb). It was working quite nicely for some time but whenever I switch branches back to master it starts to behave strangely - not properly setting breakpoints etc. I spent too much time now trying to get this extension to work than writing code for PR... 

So my question for ClickHouse devs is what tools, or extensions, are most devs at ClickHouse using for visual debug with LLDB, such that stepping through code, setting breakpoints is possible? Any personal preferences and recommendations?

**Comment by Blargian** [comment]
Edit: eventually fixed the issue with Code LLDB extension, it works well enough for debugging in VS Code. 

add to launch.json:
```
{
    "version": "0.2.0",
    "configurations": [
        {
            "type": "lldb",
            "request": "launch",
            "name": "Debug-Server",
            "program": "/home/shaun/Desktop/ClickHouse/build/programs/clickhouse",
            "args": ["server","--config-file","/home/shaun/Desktop/ClickHouse//programs/server/config.xml"],
            "cwd": "${workspaceFolder}",   
            "postRunCommands": [
                "pro hand -p yes -s no -n no SIGBUS",
                "pro hand -p yes -s no -n no SIGUSR1",
                "pro hand -p yes -s no -n no SIGUSR2",
                "pro hand -p yes -s no -n no SIGTRAP",
            ],
            "sourceMap": {".":"${workspaceFolder}"}             
        },
        {
            "type": "lldb",
            "request": "launch",
            "name": "Debug-Client",
            "program": "/home/shaun/Desktop/ClickHouse/build/programs/clickhouse",
            "args": ["client"],
            "cwd": "${workspaceFolder}"
        }
    ]
}
```

`"sourceMap": {".":"${workspaceFolder}"}` was the fix for breakpoints not being set properly.

**Reply by zhangm365** [reply]
very great! Thanks a lot.

**Reply by Blargian** [reply]
@zhangm365 I found out since then that CLion IDE works really well with ClickHouse for debugging - much better than VS code. Unfortunately it is not free, only has 30 day free trial.

**Reply by zhangm365** [reply]
I haven't used the CLion IDE yet,  how to use it, can you give me  guidances? 
Thanks a lot,  I am willing to use it.

**Reply by Blargian** [reply]
I believe it should be enough to clone ClickHouse repository and install `clang` compiler with instructions in the [developer guide](https://clickhouse.com/docs/en/development/developer-instruction) and then to download and install [CLion](https://www.jetbrains.com/clion/download/#section=linux). Once you have installed it you can open the ClickHouse folder as your CLion project and you will need to change one setting - specify CMake to use `clang` compiler rather than `gcc` by adding flag `-D CMAKE_C_COMPILER=clang-18 -D CMAKE_CXX_COMPILER=clang++-18`. After this you can run the project and it will begin building. Debug works without any further setup. 

![image](https://github.com/ClickHouse/ClickHouse/assets/41984034/2c8176db-8398-42a7-85a5-ed7b52f9273b)

**Reply by zhangm365** [reply]
@Blargian Thank you a lot. I'm very glad to meet you and work with you together.

**Reply by RinChanNOWWW** [reply]
Thanks a lot for the hint `"sourceMap": {".":"${workspaceFolder}"}`.

I found that 24.3 doesn't need this config but 24.8+ (include 24.8) does (both RelWithDebInfo).
